### DISCUSSION TITLE
Multiple request contexts

### DISCUSSION THREAD

**Discussion by Gejsi** [trpc/trpc#2368] [Ideas]
# Background
From the context [docs](https://trpc.io/docs/context):
> The createContext() function is called for each request and the result is propagated to all resolvers. You can use this to pass contextual data down to the resolvers.

The important part is that the function is called for _**each**_ request.
I think that this approach may be kinda detrimental in terms of performance.
Suppose we have an app router as such:
```js
// `createRouter()` is a helper function with the app context (check the docs)
export const appRouter = createRouter()
  .transformer(superjson)
  .merge('auth.', authRouter)
  .merge('example.', exampleRouter)
```
With this structure (thanks to `createRouter()`), we can have, for example, an instance
of the user session and an ORM client (e.g prisma) in the context of each procedure:
```js
// in this example, I'm using next-auth and prisma
interface CreateContextOptions {
  session: Session | null
}

export async function createInnerContext(opts: CreateContextOptions) {
  return {
    session: opts.session,
    prisma,
  }
}

export async function createContext(opts: CreateNextContextOptions) {
  const session = await unstable_getServerSession(
    opts.req,
    opts.res,
    authOptions
  )

  return await createInnerContext({
    session,
  })
}

type Context = inferAsyncReturnType<typeof createContext>

export const createRouter = () => router<Context>()
```

# Problem
The problem is that all procedures have to wait for the context to be created, for example, the following code:
```js
const query = trpc.useQuery(['example.all'])
const mutation = trpc.useMutation('example.add', {
  async onSuccess() {
    await utils.invalidateQueries(['example.all'])
  },
})
```
causes the following database queries (using pseudocode):
```
// query all examples
1. get session -> get user
get examples

// mutate somewhere
2. get session -> get user
mutate examples

// after the successful mutation
3. get session -> get user
get examples
```
We can notice that we need to re-query the user session for each procedure and this can really
slow things down.
IMO, the biggest problem, however, is the fact that even procedures that don't necessarily need
auth will have to wait for the user session query.

# Solution
I don't really have a solution, since I'm kind of a newbie with trpc (great project btw).

1. Is there a way to make multiple contexts in order to have different types of `createRouter()`? Maybe this approach shouldn't even be considered, thinking about the logical implications and the TypeScript/NextJs boilerplate. 
2. If the previous question makes no sense, is there a way to avoid the problem that I've encountered?

**Comment by sachinraja** [comment]
First of all, thanks for outlining the issue you ran into because I don't think you actually want multiple contexts.

You don't have to fetch the session in `createContext`. To avoid running it for every procedure on a batch call you don't want to do this in middleware either.

You can instead pass a function down and call that in your procedures.
```ts
interface CreateContextOptions {
  getSession: () => Promise<Session | null>
}

export async function createInnerContext(opts: CreateContextOptions) {
  return {
    getSession: opts.getSession,
    prisma,
  }
}

export async function createContext(opts: CreateNextContextOptions) {
  let session: Session | undefined = undefined;
  const getSession = async () => {
    // "cache" session
    if (session) return session;
    session = await unstable_getServerSession(opts.req, opts.res, authOptions);
    return session;
  };

  return await createInnerContext({
    getSession,
  });
}
```

An alternative would be to check the URL in the `createContext` and skip getting the session for certain procedure paths.

**Reply by Gejsi** [reply]
That's a great workaround! There are some changes that I would make though.
Firstly, a minor adjustment is to have the `session` as a `Session | null` rather than `Session | undefined`.

Secondly, I think there is an issue with the "caching" of the session: since you are doing it inside `createContext()`, session will alway be `undefined` since the function is called per request.
I think the solution is simply putting it outside the function, so the file as whole becomes:
```ts
import { inferAsyncReturnType, router } from '@trpc/server'
import type { CreateNextContextOptions } from '@trpc/server/adapters/next'
import {
  Session,
  unstable_getServerSession as getServerSession,
} from 'next-auth'
import { authOptions } from '../../pages/api/auth/[...nextauth]'
import { prisma } from '../db/prisma'

interface CreateContextOptions {
  getUserSession: () => Promise<Session | null>
}

/**
 * Inner function for `createContext` where we create the context.
 * This is useful for:
 * - Integration testing when we don't want to mock Next.js' request/response
 * - `createSSGHelpers` where we don't have a `req` or `res`
 */
export async function createInnerContext(opts: CreateContextOptions) {
  return {
    getUserSession: opts.getUserSession,
    prisma,
  }
}

let session: Session | null = null

const getUserSession = async (opts: CreateNextContextOptions) => {
  // if there is already a session, don't re-fetch it
  if (session) return session

  session = await getServerSession(opts.req, opts.res, authOptions)
  return session
}

export async function createContext(opts: CreateNextContextOptions) {
  return await createInnerContext({
    getUserSession: () => getUserSession(opts),
  })
}

type Context = inferAsyncReturnType<typeof createContext>

export const createRouter = () => router<Context>()

```

Thanks again.

**Reply by sachinraja** [reply]
Do not put it outside the function, that is a huge security risk. If multiple requests come in at the same time (from different users) they'll share that session.

It should work inside the `createContext` function, it's a closure.

**Reply by Gejsi** [reply]
You are absolutely right.
However, it cannot work inside `createContext` because the function is called for each request.
Going back to the first example:
```ts
const query = trpc.useQuery(['example.all'])
const mutation = trpc.useMutation('example.add', {
  async onSuccess() {
    await utils.invalidateQueries(['example.all'])
  },
})
```
The result still is the same as before, because `createContext` is called three times and `session` is always set as `null` at the start of the function, so it cannot be memoized properly.

```
// query all examples
1. get session -> get user
get examples
// mutate somewhere
2. get session -> get user
[...]
```
The procedure may look something like:
```ts
export const appRouter = createRouter()
  .transformer(superjson)
   .query('all', {
    async resolve({ ctx }) {
      // this session is always re-fetched
      const session = await ctx.getUserSession()
      return await ctx.prisma.example.findMany({
        where: { authorId: session?.user?.id },
      })
    },
  })

```

Am I missing something important? To better understand the problem I can also set up a repo or an online playground.

**Reply by sachinraja** [reply]
You're right, those are separate calls. I was solving this issue first:
> IMO, the biggest problem, however, is the fact that even procedures that don't necessarily need
auth will have to wait for the user session query.

Any framework will do multiple session calls, simply because those are multiple requests. You can possibly use Redis to do some caching. You could also look into JWTs.

**Reply by Gejsi** [reply]
Your solution, with some minor tweaks, solves the problem of passing the session to procedures that don't need it, and that's great.
I'd suggest writing it (maybe even the "checking the URL" approach) in the docs. I can contribute, if it's needed.

About:
> Any framework will do multiple session calls, simply because those are multiple requests. You can possibly use Redis to do some caching. You could also look into JWTs.

That's true, I started this discussion because I was looking for a solution for a practical use-case.
I was just playing around with tRPC and next-auth, and I noticed that every request always took more than 1.5s:
- ~ 1s to fetch the session
- ~ 400ms to fetch some random stuff (e.g. posts)

This is really bad performance wise and has nothing to do with tRPC; I'm still unable to understand how to get better results, and stuff like Redis are a bit of an overkill, since all I'm requesting is like three posts 🤷‍♂️
